featurebyte/worker/task_executor.py

```diff
@@ -61,13 +69,20 @@ def run_async(coro: Awaitable[Any]) -> Any:
         thread = Thread(target=start_background_loop, args=(loop,), daemon=True)
         thread.start()
 
-    tasks = asyncio.all_tasks()
-    logger.debug("Asyncio tasks", extra={"num_tasks": len(tasks)})
+    logger.debug("Asyncio tasks", extra={"num_tasks": len(asyncio.all_tasks(loop=loop))})
+
+    logger.info("Start task", extra={"timeout": timeout})
     future = asyncio.run_coroutine_threadsafe(coro, loop)
-    event = gevent.event.Event()
-    future.add_done_callback(lambda _: event.set())
-    event.wait()
-    return future.result()
+    try:
+        with gevent.Timeout(seconds=timeout, exception=TimeoutError):
+            event = gevent.event.Event()
+            future.add_done_callback(lambda _: event.set())
+            event.wait()
+            return future.result()
+    except TimeoutError as exc:
+        # try to cancel the job if it has not started
+        future.cancel()
+        raise SoftTimeLimitExceeded(f"Task timed out after {timeout}s") from exc
```

Comment:
Will it throw another exception (when calling `future.cancel()`) if the job is started already?

Reply:
cancel will not raise any exception but return False if the task cannot be cancelled